checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,850 @@
+        // The obligation set calculated for RLC differs from the Inference process. In the
+        // Inference process, it exclusively tracks parameters with non-empty must-call types,
+        // whether they have the @Owning annotation or not. However, there are some shared
+        // computations, such as 'updateObligationsWithInvocationResult,' which is used during
+        // inference and could potentially affect the RLC result if it were called before the
+        // checking phase. However, calling 'updateObligationsWithInvocationResult()' will not have
+        // any side effects on the outcome of the Resource Leak Checker. This is because the
+        // inference occurs within the postAnalyze method of the ResourceLeakAnnotatedTypeFactory,
+        // once the consistency analyzer has completed its process
         if (node instanceof MethodInvocationNode) {
-          checkForMustCallInvocationOnField((MethodInvocationNode) node);
+          mcca.updateObligationsWithInvocationResult(obligations, node);
+          computeOwningFromInvocation(obligations, (MethodInvocationNode) node);
+        } else if (node instanceof ObjectCreationNode) {
+          mcca.updateObligationsWithInvocationResult(obligations, node);
+        } else if (node instanceof AssignmentNode) {
+          analyzeOwnershipTransferAtAssignment(obligations, (AssignmentNode) node);
+        }
+      }
+
+      addNonExceptionalSuccessorsToWorklist(obligations, current.block, visited, worklist);
+    }
+  }
+
+  /**
+   * Returns a set of obligations representing the formal parameters of the current method that have
+   * non-empty MustCall annotations. Returns an empty set if the given CFG doesn't correspond to a
+   * method body.
+   *
+   * @param cfg the control flow graph of the method to check
+   * @return a set of obligations representing the parameters with non-empty MustCall
+   */
+  private Set<Obligation> getNonEmptyMCParams(ControlFlowGraph cfg) {
+    // TODO what about lambdas?
+    if (cfg.getUnderlyingAST().getKind() != UnderlyingAST.Kind.METHOD) {
+      return Collections.emptySet();
+    }
+    Set<Obligation> result = null;
+    for (VariableTree param : methodTree.getParameters()) {
+      if (resourceLeakAtf.declaredTypeHasMustCall(param)) {
+        VariableElement paramElement = TreeUtils.elementFromDeclaration(param);
+        if (result == null) {
+          result = new HashSet<>(2);
+        }
+        result.add(
+            new Obligation(
+                ImmutableSet.of(new ResourceAlias(new LocalVariable(paramElement), param))));
+      }
+    }
+    return result != null ? result : Collections.emptySet();
+  }
+
+  /**
+   * Returns all owning fields within the enclosing class. These are the fields that have been
+   * annotated with the {@code @Owning} annotation or inferred as owning.
+   *
+   * @return the owning fields
+   */
+  private Set<VariableElement> getOwningFields() {
```

Comment:
It's confusing that a method `getOwningFields()` mutates the `owningFields` set.  If it's essential that the update happen here, we should document and rename the method, e.g.:
```suggestion
  /**
   * Updates the owning fields set for this class to include all fields inferred as owning in this iteration.
   *
   * @return the owning fields, including fields inferred as owning from the current iteration
   */
  private Set<VariableElement> updateOwningFields() {
```